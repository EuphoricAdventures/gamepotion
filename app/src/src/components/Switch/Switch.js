import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import uuid from 'uuid'

import { font, colours } from '../../styleAbstractions'

const StyledSwitch = styled.div`
  // nearly all of this code is stolen from http://www.designcouch.com/home/why/2013/09/19/ios7-style-pure-css-toggle/
  // thank you Jesse Couch
  height: 2rem;
  > input {
    display: none;
  }
  > label {
    float: left;
    height: 2rem;
  }
  > label + label {
    line-height: 2rem;
    margin-left: 0.6666rem;
    ${font}
    font-size: 80%;
    color: #6c7a89;
  }
  > input + label {
    display: block;
    position: relative;
    // off colour background
    background-color: ${colours.back};
    // off colour border
    box-shadow: inset 0 0 0px 2px rgb(192, 192, 192);
    text-indent: -5000px;
    width: calc(2rem + (2rem * (2 / 3)));
    border-radius: 1rem;
  }
  > input + label:before {
    content: "";
    position: absolute;
    display: block;
    width: 2rem;
    height: 2rem;
    top: 0;
    left: 0;
    border-radius: 1rem;
    transition: .25s ease-in-out;
  }
  > input + label:after {
    content: "";
    position: absolute;
    display: block;
    height: 2rem;
    width: 2rem;
    top: 0;
    left: 0;
    border-radius: 1rem;
    background-color: white;
    // off colour border
    box-shadow: inset 0 0 0 2px rgb(192, 192, 192);
    transition: .25s ease-in-out;
  }
  > input:checked + label:before {
    width: calc(2rem + (2rem * (2 / 3)));
    // on colour background
    background-color: #4cd963;
  }
  > input:checked + label:after {
    left: calc(2rem * (2 / 3));
    // on colour nub
    background-color: white;
    box-shadow: inset 0 0 0 2px rgb(192, 192, 192);
  }
`

const Switch = ({ children, checked, onChange }) => {
  const id = uuid()

  const handleOnChange = () => {
    return onChange(!checked)
  }

  return (
    <StyledSwitch className='component--switch'>
      <input type='checkbox' id={id} checked={checked} onChange={handleOnChange} />
      <label htmlFor={id} />
      <label htmlFor={id} className='label'>{children}</label>
    </StyledSwitch>
  )
}

Switch.propTypes = {
  checked: PropTypes.bool.isRequired,
  onChange: PropTypes.func
}

Switch.defaultProps = {
  checked: false,
  onChange: () => {}
}

export default Switch
