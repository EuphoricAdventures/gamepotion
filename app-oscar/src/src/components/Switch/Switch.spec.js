import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import Switch from './Switch'

jest.mock('uuid', () => jest.fn().mockReturnValue('id'))

it('renders checked', () => {
  const tree = renderer
    .create(
      <Switch checked>Switch (checked)</Switch>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders unchecked', () => {
  const tree = renderer
    .create(
      <Switch checked={false}>Switch (unchecked)</Switch>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
