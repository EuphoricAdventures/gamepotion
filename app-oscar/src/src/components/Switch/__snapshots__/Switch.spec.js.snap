// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders checked 1`] = `
.c0 {
  height: 2rem;
}

.c0 > input {
  display: none;
}

.c0 > label {
  float: left;
  height: 2rem;
}

.c0 > label + label {
  line-height: 2rem;
  margin-left: 0.6666rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  font-size: 80%;
  color: #6c7a89;
}

.c0 > input + label {
  display: block;
  position: relative;
  background-color: #ecf0f1;
  box-shadow: inset 0 0 0px 2px rgb(192,192,192);
  text-indent: -5000px;
  width: calc(2rem + (2rem * (2 / 3)));
  border-radius: 1rem;
}

.c0 > input + label:before {
  content: "";
  position: absolute;
  display: block;
  width: 2rem;
  height: 2rem;
  top: 0;
  left: 0;
  border-radius: 1rem;
  -webkit-transition: .25s ease-in-out;
  transition: .25s ease-in-out;
}

.c0 > input + label:after {
  content: "";
  position: absolute;
  display: block;
  height: 2rem;
  width: 2rem;
  top: 0;
  left: 0;
  border-radius: 1rem;
  background-color: white;
  box-shadow: inset 0 0 0 2px rgb(192,192,192);
  -webkit-transition: .25s ease-in-out;
  transition: .25s ease-in-out;
}

.c0 > input:checked + label:before {
  width: calc(2rem + (2rem * (2 / 3)));
  background-color: #4cd963;
}

.c0 > input:checked + label:after {
  left: calc(2rem * (2 / 3));
  background-color: white;
  box-shadow: inset 0 0 0 2px rgb(192,192,192);
}

<div
  className="component--switch c0"
>
  <input
    checked={true}
    id="id"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="id"
  />
  <label
    className="label"
    htmlFor="id"
  >
    Switch (checked)
  </label>
</div>
`;

exports[`renders unchecked 1`] = `
.c0 {
  height: 2rem;
}

.c0 > input {
  display: none;
}

.c0 > label {
  float: left;
  height: 2rem;
}

.c0 > label + label {
  line-height: 2rem;
  margin-left: 0.6666rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  font-size: 80%;
  color: #6c7a89;
}

.c0 > input + label {
  display: block;
  position: relative;
  background-color: #ecf0f1;
  box-shadow: inset 0 0 0px 2px rgb(192,192,192);
  text-indent: -5000px;
  width: calc(2rem + (2rem * (2 / 3)));
  border-radius: 1rem;
}

.c0 > input + label:before {
  content: "";
  position: absolute;
  display: block;
  width: 2rem;
  height: 2rem;
  top: 0;
  left: 0;
  border-radius: 1rem;
  -webkit-transition: .25s ease-in-out;
  transition: .25s ease-in-out;
}

.c0 > input + label:after {
  content: "";
  position: absolute;
  display: block;
  height: 2rem;
  width: 2rem;
  top: 0;
  left: 0;
  border-radius: 1rem;
  background-color: white;
  box-shadow: inset 0 0 0 2px rgb(192,192,192);
  -webkit-transition: .25s ease-in-out;
  transition: .25s ease-in-out;
}

.c0 > input:checked + label:before {
  width: calc(2rem + (2rem * (2 / 3)));
  background-color: #4cd963;
}

.c0 > input:checked + label:after {
  left: calc(2rem * (2 / 3));
  background-color: white;
  box-shadow: inset 0 0 0 2px rgb(192,192,192);
}

<div
  className="component--switch c0"
>
  <input
    checked={false}
    id="id"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="id"
  />
  <label
    className="label"
    htmlFor="id"
  >
    Switch (unchecked)
  </label>
</div>
`;
